### PR TITLE
[REG-1425] Handle replay data / screenshot / validation disk buffering

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Asynchronously writes replay data, validations, screenshots, etc to disk on each tick to avoid filling up memory with the data
  - Tracks validation summary in a running fashion for each client in memory
- Uploads the data from the files directly without fully reading into memory first

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

File: Runtime/Scripts/RGBotLocalRuntime/RGBotRuntimeManager.cs
Comment (analogrelay) [inline Runtime/Scripts/RGBotLocalRuntime/RGBotRuntimeManager.cs:96]: ⚔️

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (analogrelay) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:112]: Are we guaranteed that this isn't being called simultaneously for the _same `clientId` value_? The `List<Task>` you're working with isn't thread-safe, but that should be alright as long as we're only processing one `SaveReplayDataInfo` for each `clientId` at a time.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (analogrelay) [inline Runtime/Scripts/DataCollection/RGDataCollection.cs:148]: This doesn't provide a guarantee of ordering. If tick 1 calls `SaveReplayData` and queues a write to `rgbot_validations.jsonl` and tick 2 does the same, there's no guarantee that tick 1's write finishes before tick 2's write starts. The `AppendAllTextAsync` method opens the file and locks it exclusively, so a simultaneous call to `AppendAllTextAsync` for the same file would throw because it can't acquire a lock on the file. If the writes can complete before the next tick you'd be OK, but we shouldn't really rely on it.

If you want to ensure ordering, what you need is a "Task Queue". Basically, you build up a chain of `.ContinueWith` calls that ensure that task `N` doesn't **start** until task `N-1` completes . It's written something like this:

```csharp
private object _lock = new object();
private Task _activeTask = Task.CompletedTask;

public void QueueSomething()
{
    async Task DoTheWork()
    {
        // This actually does the work.
        // Because of the queue logic, it's guaranteed to complete fully before it gets called again.
    }

    // If you're confident QueueSomething can't be called simultaneously, you can omit the lock
    // Problems arise if two threads are messing with _activeTask at once.
    lock(_lock)
    {
        // Note that we _replace_ "_activeTask"!
        // This means that _activeTask now represents the completion of _all_ queued work so far
        // So, when the next call to QueueSomething comes in, it's not going to _start_ it's work until the entire queue up to that point completes.
        _activeTask = _activeTask.ContinueWith(_ => DoTheWork());
    }
}
```

That's a rough version. It'll have a bunch of allocations of `Task`s though, so it might not be the best approach.

Another approach would be to enqueue the write request with all the validations to a static `ConcurrentQueue` somewhere and have a single background thread grab write requests in order and flush them to disk.

Comment (analogrelay) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:148]: Another solution would be to buffer the validations in separate files for each tick. Then you'd end up in the same place as replay data, which doesn't have conflicting writes to the same file, but does still produce a lot of async work to be tracked and waited for.

**Review by analogrelay — CHANGES_REQUESTED** [review]
Looks mostly good, but I do see a race condition that may result in errors appending validations to the file.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:112]: Good callout, but yes we are guaranteed not to call this in parallel for the same clientId.  This is called from each client botrunner thread in a manner that blocks said thread until this completes.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:148]: eesh.. it looks like our `unity.js` version of this code probably has a similar issue

For this C# case, I'm going to go with a simple solution for now.

The code will wait for the prior write to finish on the next call.  Normally that async write will complete before the next tick comes in, so this shouldn't hold up our client processing.  If we find that we're having issues holding up the client tick processing with high tick rates or other cases we can pursue a solution more in line with your proposal above.

File: Runtime/Scripts/DataCollection/RGDataCollection.cs
Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:150]: Ah I see @analogrelay may have left a similar review

Comment (RG-nAmKcAz) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:150]: see about 6 lines above your question where it waits for the prior one to finish

Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:150]: But doesn't await just wait for the task to finish, nothing to say that the tasks are still running in parallel? Maybe this is a C# thing I'm not familiar with (im thinking mostly in Python), but don't the tasks get started right when you create them? In other words, by the time we get to that await line, you may be waiting for one task to finish, but other tasks can still be running at the same time?

Comment (vontell) [reply Runtime/Scripts/DataCollection/RGDataCollection.cs:150]: Oh... I'm reading the wrong task. I thought this was a parallel task to write into JSONL for each validation, ignore what I am saying!

**Review by vontell — APPROVED** [review]

**Review by addisonbgross — APPROVED** [review]
some smooth async work 🧊

**Comment by RG-nAmKcAz** [comment]
@analogrelay Notifying you that this is waiting on your second review/approval

**Pull request merged**
